library/net.cpp

```diff
@@ -1424,3 +1433,274 @@ int dnet_trans_forward(struct dnet_io_req *r, struct dnet_net_state *orig, struc
+
+		auto r = static_cast<dnet_io_req *>(calloc(1, sizeof(dnet_io_req)));
+		if (!r)
+			return -ENOMEM;
+
+		r->io_req_type = DNET_IO_REQ_TYPED_RESPONSE;
+		r->response_info = response_info.release();
+
+		r->st = dnet_state_get(st);
+		dnet_schedule_io(st->n, r);
+
+		return 0;
+	};
+
+	n2_repliers repliers_wrappers;
+
+	repliers_wrappers.on_reply_error = [on_reply_error = std::move(repliers.on_reply_error),
+                                            enqueue_response](int errc) -> int {
+		return enqueue_response(std::bind(on_reply_error, errc));
+	};
+
+	repliers_wrappers.on_reply = [on_reply = std::move(repliers.on_reply),
+	                              enqueue_response](std::unique_ptr<n2_message> msg) -> int {
+		return enqueue_response(n2_bind_noncopyable(on_reply, std::move(msg)));
+	};
+
+	return repliers_wrappers;
+}
+
+int n2_complete_trans_via_response_holder(dnet_trans *t, n2_response_info *response_info) {
+	return c_exception_guard(response_info->response_holder, t->st->n, __FUNCTION__);
+}
+
+// TODO(sabramkin): Try rework to n2_trans_alloc_send. In new mechanic we don't need to separate alloc and send
+static int n2_trans_send(dnet_trans *t, n2_request_info *request_info) {
+	using namespace ioremap::elliptics;
+
+	struct dnet_net_state *st = t->st;
+	struct dnet_test_settings test_settings;
+	int err;
+
+	dnet_trans_get(t);
+
+	BOOST_SCOPE_EXIT(&t) {
+		dnet_trans_put(t);
+	} BOOST_SCOPE_EXIT_END
+
+	pthread_mutex_lock(&st->trans_lock);
+	err = dnet_trans_insert_nolock(st, t);
+	if (!err) {
+		dnet_trans_update_timestamp(t);
+		dnet_trans_insert_timer_nolock(st, t);
+	}
+	pthread_mutex_unlock(&st->trans_lock);
+	if (err)
+		return err;
+
+	if (t->n->test_settings && !dnet_node_get_test_settings(t->n, &test_settings) &&
+	    test_settings.commands_mask & (1 << t->command))
+		return err;
+
+	auto repliers_wrappers = n2_make_repliers_via_request_queue(st,
+	                                                            request_info->request->cmd,
+	                                                            std::move(request_info->repliers));
+
+	n2::net_state_get_protocol(st)->send_request(st,
+	                                             std::move(request_info->request),
+	                                             std::move(repliers_wrappers));
+
+	return err;
+}
+
+int n2_trans_forward(n2_request_info *request_info, struct dnet_net_state *orig, struct dnet_net_state *forward) {
+	auto deadline = request_info->request->deadline;
+	timespec wait_ts = dnet_time_is_empty(&deadline)
+		? orig->n->wait_ts
+		: dnet_time_left_to_timeout(deadline);
+
+	if (!wait_ts.tv_sec && !wait_ts.tv_nsec) {
+		return -ETIMEDOUT;
+	}
+
+	auto t = dnet_trans_alloc(orig->n, 0);
+	if (!t)
+		return -ENOMEM;
+
+	dnet_cmd *cmd = &request_info->request->cmd;
+
+	t->rcv_trans = cmd->trans; // TODO(sabramkin): Is it necessary in new mechanic?
+	t->trans = cmd->trans = atomic_inc(&orig->n->trans);
```

Comment:
So only `request_info->request->cmd->trans` is changed. What about `request_info->cmd->trans`?

Reply:
Forgot it =(